File: src-tauri/src/embedded.rs
use std::fs;
use std::io::{Cursor, Read};
use std::path::{Path, PathBuf};

const FLOWSEAL_SNAPSHOT: &[u8] = include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/resources/flowseal-main.zip"));

const FLOWSEAL_ENGINE: &[u8] = include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/resources/engine-flowseal.zip"));

pub const SNAPSHOT_INFO: &str = "Official Flowseal main snapshot bundled at build time";
/// Версия вшитого движка Flowseal (входит в состав exe).
pub const ENGINE_VERSION: &str = "1.10.2";
pub const ENGINE_INFO: &str = "Flowseal 1.10.2 release archive bundled in the executable";

/// Распаковывает встроенный релиз движка Flowseal в `<data>/engines/flowseal`,
/// если там ещё нет exe. Возвращает путь к корню движка, если он готов.
pub fn ensure_embedded_engine(data: &Path) -> Result<Option<PathBuf>, String> {
    let root = data.join("engines").join(crate::config::ENGINE_FLOWSEAL);
    if let Some(existing) = engine_root_for(&root) {
        return Ok(Some(existing));
    }
    fs::create_dir_all(&root).map_err(|e| e.to_string())?;
    extract_all(FLOWSEAL_ENGINE, &root)?;
    let found = match engine_root_for(&root) {
        Some(found) => found,
        None => return Err("во встроенном архиве flowseal не найден winws.exe".into()),
    };
    neutralize_author_autoupdate(&found);
    Ok(Some(found))
}

/// Отключает авторскую авто-проверку обновлений Flowseal: `utils/check_updates.enabled`
/// заставляет каждый *.bat вызывать `service.bat check_updates`, который открывает
/// страницу релиза в браузере. Нам это не нужно (обновления ведёт наш GUI).
pub fn neutralize_author_autoupdate(root: &Path) {
    let flag = root.join("utils").join("check_updates.enabled");
    if flag.exists() {
        let _ = fs::rename(&flag, root.join("utils").join("check_updates.enabled.zgui_disabled"));
        if flag.exists() {
            let _ = fs::remove_file(&flag);
        }
    }
}

/// Публичная обёртка для нормализации корня уже распакованного движка.
pub fn engine_root_for_public(dir: &Path) -> Option<PathBuf> {
    engine_root_for(dir)
}

/// Находит реальный корень движка внутри распакованной папки Flowseal —
/// каталог с `bin/winws.exe`.
fn engine_root_for(dir: &Path) -> Option<PathBuf> {
    let rel = crate::config::find_exe(dir, "winws.exe")?;
    let exe_path = dir.join(rel.replace('/', std::path::MAIN_SEPARATOR_STR));
    // Поднимаемся от exe вверх до каталога, содержащего marker (bin).
    let mut cur = exe_path.parent().map(|p| p.to_path_buf())?;
    loop {
        if cur.join("bin").is_dir() {
            return Some(cur);
        }
        match cur.parent() {
            Some(p) if p.starts_with(dir) => cur = p.to_path_buf(),
            _ => break,
        }
    }
    // Fallback — сам каталог.
    Some(dir.to_path_buf())
}

/// Полная распаковка архива с обрезкой общего верхнего каталога (безопасно от zip-slip).
fn extract_all(bytes: &[u8], dest: &Path) -> Result<usize, String> {
    let mut archive = zip::ZipArchive::new(Cursor::new(bytes)).map_err(|e| e.to_string())?;

    let mut top: Option<String> = None;
    let mut uniform = true;
    for i in 0..archive.len() {
        let name = archive.by_index(i).map_err(|e| e.to_string())?.name().to_string();
        let trimmed = name.trim_end_matches('/');
        if trimmed.is_empty() {
            continue;
        }
        match trimmed.split_once('/') {
            // Запись внутри каталога: проверяем общий верхний каталог.
            Some((first, _)) => match &top {
                None => top = Some(first.to_string()),
                Some(t) if t == first => {}
                Some(_) => {
                    uniform = false;
                    break;
                }
            },
            // Запись самого каталога-обёртки нередко идёт до его содержимого.
            // Не считаем её отдельным верхним каталогом.
            None => {
                if let Some(top) = top.as_deref() {
                    if top != trimmed {
                        uniform = false;
                    }
                }
            }
        }
    }
    let strip = if uniform { top } else { None };

    let mut written = 0;
    for i in 0..archive.len() {
        let mut entry = archive.by_index(i).map_err(|e| e.to_string())?;
        let is_dir = entry.is_dir();
        let name = entry.name().to_string();
        let mut relative = name.clone();
        if let Some(t) = &strip {
            relative = relative
                .strip_prefix(&format!("{}/", t))
                .map(|s| s.to_string())
                .unwrap_or(relative);
        }
        if relative.trim_end_matches('/').is_empty() {
            continue;
        }
        let Some(rel) = safe_relative(relative.trim_end_matches('/')) else { continue };
        let out = dest.join(&rel);
        if !out.starts_with(dest) {
            continue;
        }
        if is_dir {
            let _ = fs::create_dir_all(&out);
            continue;
        }
        if let Some(parent) = out.parent() {
            let _ = fs::create_dir_all(parent);
        }
        let mut content = Vec::with_capacity(entry.size() as usize);
        entry.read_to_end(&mut content).map_err(|e| e.to_string())?;
        crate::config::atomic_write(&out, &content)?;
        written += 1;
    }
    Ok(written)
}

pub fn seed_catalog(data: &Path) -> Result<usize, String> {
    let mut written = 0;
    // Remove the legacy system-hosts catalog entry from older portable data.
    let legacy_hosts = data.join("catalog/flowseal/.service/hosts");
    if legacy_hosts.exists() {
        let _ = fs::remove_file(legacy_hosts);
    }
    written += extract_missing(FLOWSEAL_SNAPSHOT, data, flowseal_destination)?;

    let info = data.join("catalog/source-info.txt");
    if !info.exists() {
        crate::config::atomic_write(&info, SNAPSHOT_INFO.as_bytes())?;
        written += 1;
    }
    Ok(written)
}

fn extract_missing(
    bytes: &[u8],
    data: &Path,
    destination: fn(&str) -> Option<PathBuf>,
) -> Result<usize, String> {
    let mut archive = zip::ZipArchive::new(Cursor::new(bytes)).map_err(|e| e.to_string())?;
    let mut written = 0;

    for i in 0..archive.len() {
        let mut entry = archive.by_index(i).map_err(|e| e.to_string())?;
        if entry.is_dir() {
            continue;
        }
        let name = entry.name().to_string();
        let Some(relative) = name.split_once('/').map(|(_, tail)| tail) else {
            continue;
        };
        let Some(destination) = destination(relative) else {
            continue;
        };
        let destination = data.join(destination);
        if destination.exists() {
            continue;
        }

        let mut content = Vec::with_capacity(entry.size() as usize);
        entry.read_to_end(&mut content).map_err(|e| e.to_string())?;
        crate::config::atomic_write(&destination, &content)?;
        written += 1;
    }
    Ok(written)
}

fn flowseal_destination(relative: &str) -> Option<PathBuf> {
    match relative {
        "LICENSE.txt" => return Some(PathBuf::from("catalog/sources/flowseal/LICENSE.txt")),
        "README.md" => return Some(PathBuf::from("catalog/sources/flowseal/README.md")),
        "service.bat" => return Some(PathBuf::from("catalog/sources/flowseal/service.bat")),
        _ => {}
    }

    if let Some(name) = relative.strip_prefix(".service/") {
        if matches!(name, "version.txt" | "ipset-service.txt") {
            return safe_relative(name).map(|p| PathBuf::from("catalog/flowseal/.service").join(p));
        }
    }
    if let Some(name) = relative.strip_prefix("lists/") {
        if matches!(
            name,
            "list-general.txt" | "list-google.txt" | "list-exclude.txt" | "ipset-exclude.txt" | "ipset-all.txt"
        ) {
            return safe_relative(name).map(|p| PathBuf::from("catalog/flowseal/lists").join(p));
        }
    }
    if !relative.contains('/') && relative.to_ascii_lowercase().ends_with(".bat") && !relative.eq_ignore_ascii_case("service.bat") {
        return safe_relative(relative).map(|p| PathBuf::from("catalog/flowseal/raw").join(p));
    }
    None
}

fn safe_relative(path: &str) -> Option<PathBuf> {
    let mut out = PathBuf::new();
    for part in path.split('/') {
        if part.is_empty() || part == "." {
            continue;
        }
        if part == ".." || part.contains('\\') || part.contains(':') {
            return None;
        }
        out.push(part);
    }
    (!out.as_os_str().is_empty()).then_some(out)
}

pub fn copy_missing(source: &Path, destination: &Path) {
    if source.is_file() && !destination.exists() {
        if let Some(parent) = destination.parent() {
            let _ = fs::create_dir_all(parent);
        }
        let _ = fs::copy(source, destination);
    }
}

pub fn copy_tree_missing(source: &Path, destination: &Path) {
    let Ok(entries) = fs::read_dir(source) else {
        return;
    };
    for entry in entries.flatten() {
        let src = entry.path();
        let dst = destination.join(entry.file_name());
        if src.is_dir() {
            copy_tree_missing(&src, &dst);
        } else {
            copy_missing(&src, &dst);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn seeds_the_official_config_catalog() {
        let data = std::env::temp_dir().join(format!(
            "zgui-embedded-test-{}",
            std::process::id()
        ));
        let _ = fs::remove_dir_all(&data);
        fs::create_dir_all(&data).unwrap();

        let written = seed_catalog(&data).unwrap();

        assert!(written > 0);
        assert!(data.join("catalog/flowseal/raw/general.bat").is_file());
        assert!(data.join("catalog/flowseal/lists/list-general.txt").is_file());
        assert!(!data.join("catalog/flowseal/.service/hosts").exists());

        let _ = fs::remove_dir_all(&data);
    }

    #[test]
    fn embedded_engines_contain_exe() {
        let data = std::env::temp_dir().join(format!("zgui-engine-test-{}", std::process::id()));
        let _ = fs::remove_dir_all(&data);
        fs::create_dir_all(&data).unwrap();

        let fs_root = ensure_embedded_engine(&data).unwrap();
        assert!(fs_root.is_some(), "встроенный flowseal не распаковался");
        let fs_root = fs_root.unwrap();
        assert!(crate::config::find_exe(&fs_root, "winws.exe").is_some());

        // Корень должен содержать bin/, иначе лаунчер соберёт неверные пути
        // (регрессия «process exited immediately»).
        assert!(fs_root.join("bin").is_dir(), "flowseal root без bin/: {:?}", fs_root);

        let _ = fs::remove_dir_all(&data);
    }

    #[test]
    fn disables_author_autoupdate_flag() {
        let data = std::env::temp_dir().join(format!("zgui-au-{}", std::process::id()));
        let _ = fs::remove_dir_all(&data);
        let utils = data.join("utils");
        fs::create_dir_all(&utils).unwrap();
        let flag = utils.join("check_updates.enabled");
        fs::write(&flag, b"1").unwrap();

        neutralize_author_autoupdate(&data);

        assert!(!flag.exists(), "флаг автопроверки должен быть отключён");
        assert!(utils.join("check_updates.enabled.zgui_disabled").exists());
        let _ = fs::remove_dir_all(&data);
    }
}

File: winforms/src/ZapretGui.Core/Embedded/Embedded.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;
using System.Text;
using ZapretGui.Core.Config;
using ZapretGui.Core.Log;
using ZapretGui.Core.Runtime;
using ZapretGui.Core.Util;

namespace ZapretGui.Core.Embedded
{
    /// <summary>
    /// Встроенные в исполняемый файл архивы Flowseal: движок (engine-flowseal.zip)
    /// и каталог стратегий (flowseal-main.zip) (embedded.rs).
    /// </summary>
    public static class Embedded
    {
        /// <summary>Версия вшитого движка Flowseal (embedded.rs:11).</summary>
        public const string EngineVersion = "1.10.2";

        /// <summary>Описание вшитого движка для UI (embedded.rs:12).</summary>
        public const string EngineInfo = "Flowseal 1.10.2 release archive bundled in the executable";

        /// <summary>Метка источника каталога (embedded.rs:9).</summary>
        public const string SnapshotInfo = "Official Flowseal main snapshot bundled at build time";

        /// <summary>Байты engine-flowseal.zip из ресурсов программы.</summary>
        public static byte[] EngineZip;

        /// <summary>Байты flowseal-main.zip из ресурсов программы.</summary>
        public static byte[] SnapshotZip;

        /// <summary>
        /// Распаковывает встроенный релиз движка Flowseal в
        /// &lt;data&gt;/engines/flowseal, если там ещё нет exe. Возвращает путь к
        /// корню движка или null (embedded.rs:16-29).
        /// </summary>
        public static string EnsureEmbeddedEngine(string dataDir)
        {
            var root = Path.Combine(dataDir, "engines", Engines.Flowseal);
            var existing = EngineRootFor(root);
            if (existing != null) { return existing; }

            if (EngineZip == null)
            {
                LogRing.Write("err", "embedded", "встроенный архив движка не подключён");
                return null;
            }

            try { Directory.CreateDirectory(root); }
            catch (Exception e) { LogRing.Write("err", "embedded", "не создать каталог движка: " + e.Message); return null; }

            try { ExtractAll(EngineZip, root); }
            catch (Exception e) { LogRing.Write("err", "embedded", "распаковка движка не удалась: " + e.Message); return null; }

            var found = EngineRootFor(root);
            if (found == null)
            {
                LogRing.Write("err", "embedded", "во встроенном архиве flowseal не найден winws.exe");
                return null;
            }
            NeutralizeAuthorAutoupdate(found);
            return found;
        }

        /// <summary>
        /// Отключает авторскую авто-проверку обновлений Flowseal:
        /// utils/check_updates.enabled заставляет каждый *.bat вызывать
        /// service.bat check_updates, который открывает страницу релиза в
        /// браузере. Нам это не нужно — обновления ведёт GUI (embedded.rs:34-42).
        /// </summary>
        public static void NeutralizeAuthorAutoupdate(string root)
        {
            var flag = Path.Combine(root, "utils", "check_updates.enabled");
            if (!File.Exists(flag)) { return; }
            var disabled = flag + ".zgui_disabled";
            try
            {
                if (File.Exists(disabled)) { File.Delete(disabled); }
                File.Move(flag, disabled);
            }
            catch { }
            if (File.Exists(flag))
            {
                try { File.Delete(flag); } catch { }
            }
        }

        /// <summary>Нормализованный корень уже распакованного движка (embedded.rs:51-67).</summary>
        public static string EngineRootFor(string dir)
        {
            if (!Directory.Exists(dir)) { return null; }
            var rel = Processes.FindExe(dir, Engines.WinwsExe);
            if (rel == null) { return null; }

            var exePath = Path.Combine(dir, rel.Replace('/', Path.DirectorySeparatorChar));
            var cur = Path.GetDirectoryName(exePath);
            while (cur != null && IsInside(dir, cur))
            {
                if (Directory.Exists(Path.Combine(cur, "bin"))) { return cur; }
                var parent = Path.GetDirectoryName(cur);
                if (parent == null || !IsInside(dir, parent)) { break; }
                cur = parent;
            }
            return dir;
        }

        /// <summary>
        /// Полная распаковка архива с обрезкой общего верхнего каталога
        /// и защитой от zip-slip (embedded.rs:70-137). Возвращает число файлов.
        /// </summary>
        public static int ExtractAll(byte[] bytes, string dest)
        {
            Directory.CreateDirectory(dest);
            string strip;
            using (var probe = new ZipArchive(new MemoryStream(bytes), ZipArchiveMode.Read))
            {
                strip = DetectTopDir(probe);
            }

            var written = 0;
            using (var archive = new ZipArchive(new MemoryStream(bytes), ZipArchiveMode.Read))
            {
                foreach (var entry in archive.Entries)
                {
                    var name = entry.FullName;
                    var relative = name;
                    if (!string.IsNullOrEmpty(strip))
                    {
                        var prefix = strip + "/";
                        if (relative.StartsWith(prefix, StringComparison.Ordinal))
                        {
                            relative = relative.Substring(prefix.Length);
                        }
                    }
                    var trimmed = relative.TrimEnd('/');
                    if (trimmed.Length == 0) { continue; }

                    var rel = SafeRelative(trimmed);
                    if (rel == null) { continue; }
                    var outPath = Path.Combine(dest, rel);
                    if (!IsInside(dest, outPath)) { continue; }

                    var isDir = name.EndsWith("/");
                    if (isDir)
                    {
                        try { Directory.CreateDirectory(outPath); } catch { }
                        continue;
                    }

                    try
                    {
                        var parent = Path.GetDirectoryName(outPath);
                        if (!string.IsNullOrEmpty(parent)) { Directory.CreateDirectory(parent); }
                        using (var src = entry.Open())
                        using (var buf = new MemoryStream())
                        {
                            src.CopyTo(buf);
                            Text.AtomicWrite(outPath, buf.ToArray());
                        }
                        written++;
                    }
                    catch { }
                }
            }
            return written;
        }

        /// <summary>
        /// Seeds каталог стратегий из встроенного снимка Flowseal: копирует только
        /// недостающие файлы (embedded.rs:139-154 + 156-214).
        /// </summary>
        public static int SeedCatalog(string dataDir)
        {
            var written = 0;
            // Удаляем устаревший системный hosts из старых портативных данных.
            var legacyHosts = Path.Combine(dataDir, "catalog", "flowseal", ".service", "hosts");
            if (File.Exists(legacyHosts))
            {
                try { File.Delete(legacyHosts); } catch { }
            }

            if (SnapshotZip != null)
            {
                written += ExtractMissing(SnapshotZip, dataDir);
            }

            var info = Path.Combine(dataDir, "catalog", "source-info.txt");
            if (!File.Exists(info))
            {
                try
                {
                    Text.AtomicWrite(info, Encoding.UTF8.GetBytes(SnapshotInfo));
                    written++;
                }
                catch { }
            }
            return written;
        }

        // Распаковка только недостающих файлов каталога (embedded.rs:156-187).
        private static int ExtractMissing(byte[] bytes, string dataDir)
        {
            var written = 0;
            using (var archive = new ZipArchive(new MemoryStream(bytes), ZipArchiveMode.Read))
            {
                foreach (var entry in archive.Entries)
                {
                    if (entry.FullName.EndsWith("/")) { continue; }
                    var slash = entry.FullName.IndexOf('/');
                    if (slash < 0) { continue; }
                    var relative = entry.FullName.Substring(slash + 1);
                    if (relative.Length == 0) { continue; }

                    var destination = FlowsealDestination(relative);
                    if (destination == null) { continue; }
                    var target = Path.Combine(dataDir, destination);
                    if (File.Exists(target)) { continue; }

                    try
                    {
                        var parent = Path.GetDirectoryName(target);
                        if (!string.IsNullOrEmpty(parent)) { Directory.CreateDirectory(parent); }
                        using (var src = entry.Open())
                        using (var buf = new MemoryStream())
                        {
                            src.CopyTo(buf);
                            Text.AtomicWrite(target, buf.ToArray());
                        }
                        written++;
                    }
                    catch { }
                }
            }
            return written;
        }

        // Соответствие файлов снимка каталогу программы (embedded.rs:189-214).
        private static string FlowsealDestination(string relative)
        {
            switch (relative)
            {
                case "LICENSE.txt": return "catalog/sources/flowseal/LICENSE.txt";
                case "README.md": return "catalog/sources/flowseal/README.md";
                case "service.bat": return "catalog/sources/flowseal/service.bat";
            }

            string name;
            if (TryStripPrefix(relative, ".service/", out name))
            {
                if (name == "version.txt" || name == "ipset-service.txt")
                {
                    var safe = SafeRelative(name);
                    return safe == null ? null : Path.Combine("catalog/flowseal/.service", safe);
                }
            }
            if (TryStripPrefix(relative, "lists/", out name))
            {
                if (name == "list-general.txt" || name == "list-google.txt" ||
                    name == "list-exclude.txt" || name == "ipset-exclude.txt" ||
                    name == "ipset-all.txt")
                {
                    var safe = SafeRelative(name);
                    return safe == null ? null : Path.Combine("catalog/flowseal/lists", safe);
                }
            }
            if (relative.IndexOf('/') < 0 &&
                relative.EndsWith(".bat", StringComparison.OrdinalIgnoreCase) &&
                !relative.Equals("service.bat", StringComparison.OrdinalIgnoreCase))
            {
                var safe = SafeRelative(relative);
                return safe == null ? null : Path.Combine("catalog/flowseal/raw", safe);
            }
            return null;
        }

        private static bool TryStripPrefix(string text, string prefix, out string rest)
        {
            if (text.StartsWith(prefix, StringComparison.Ordinal))
            {
                rest = text.Substring(prefix.Length);
                return true;
            }
            rest = null;
            return false;
        }

        /// <summary>
        /// Безопасный относительный путь: отсекает .., абсолютные пути и
        /// разделители Windows (embedded.rs:216-228).
        /// </summary>
        public static string SafeRelative(string path)
        {
            var parts = path.Split('/');
            var outPath = new StringBuilder();
            foreach (var part in parts)
            {
                if (part.Length == 0 || part == ".") { continue; }
                if (part == ".." || part.Contains("\\") || part.Contains(":")) { return null; }
                if (outPath.Length > 0) { outPath.Append('/'); }
                outPath.Append(part);
            }
            return outPath.Length == 0 ? null : outPath.ToString();
        }

        /// <summary>Копирует файл, только если назначения ещё нет (embedded.rs:230-237).</summary>
        public static void CopyMissing(string source, string destination)
        {
            if (!File.Exists(source) || File.Exists(destination)) { return; }
            try
            {
                var parent = Path.GetDirectoryName(destination);
                if (!string.IsNullOrEmpty(parent)) { Directory.CreateDirectory(parent); }
                File.Copy(source, destination);
            }
            catch { }
        }

        /// <summary>Копирует дерево, пропуская существующие файлы (embedded.rs:239-252).</summary>
        public static void CopyTreeMissing(string source, string destination)
        {
            string[] entries;
            try { entries = Directory.GetFileSystemEntries(source); }
            catch { return; }
            Directory.CreateDirectory(destination);
            foreach (var src in entries)
            {
                var dst = Path.Combine(destination, Path.GetFileName(src));
                if (Directory.Exists(src))
                {
                    CopyTreeMissing(src, dst);
                }
                else
                {
                    CopyMissing(src, dst);
                }
            }
        }

        // Определяет общий верхний каталог архива для обрезки (embedded.rs:73-102).
        private static string DetectTopDir(ZipArchive archive)
        {
            string top = null;
            var uniform = true;
            foreach (var entry in archive.Entries)
            {
                var trimmed = entry.FullName.TrimEnd('/');
                if (trimmed.Length == 0) { continue; }
                var slash = trimmed.IndexOf('/');
                if (slash >= 0)
                {
                    var first = trimmed.Substring(0, slash);
                    if (top == null) { top = first; }
                    else if (top != first) { uniform = false; break; }
                }
                else if (top != null && top != trimmed)
                {
                    uniform = false;
                }
            }
            return uniform ? top : null;
        }

        // path находится внутри root (сравнение по компонентам, как Path::starts_with).
        private static bool IsInside(string root, string path)
        {
            try
            {
                var fullRoot = Path.GetFullPath(root).TrimEnd('\\') + "\\";
                return Path.GetFullPath(path).StartsWith(fullRoot, StringComparison.OrdinalIgnoreCase);
            }
            catch { return false; }
        }
    }
}
